### PR TITLE
add example of re-act pattern

### PR DESCRIPTION
**Pull request by tobi**
Example implementing a version of the reason-act pattern. It's a neat example of what you can do with the 7B model, at least the instruction tuned versions. [Paper](https://arxiv.org/abs/2210.03629).

Adds `examples/reason-act.sh` and `prompts/reason-act.txt`. 

image

### CODE REVIEW

**Comment by anzz1** [comment]
One thing to try is `-r "Question:" -r "Observation:" --in-prefix " "` , since indeed the space can make it a different token.

For example, "Yes", " Yes" and "Yes " could all be different tokens. And combinations like "Yes "+"Sir" , or "Yes"+" "+"Sir", "Yes"+" Sir" and so on. Unless specified otherwise, spaces are just another byte/character like everything else. I'm not sure whether spaces had special treatment in llama's training but there definitely are spaced tokens like that.

Related (ancient) paper: [Improving Tokenisation by Alternative Treatment of Spaces](https://eprints.whiterose.ac.uk/185969/1/2204.04058v1.pdf)

File: examples/reason-act.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+
+#!/bin/bash
+
+cd `dirname $0`
+cd ..
+
+# get -m model parameter otherwise defer to default
+if [ "$1" == "-m" ]; then
+  MODEL="-m $1"
```
Comment (j-f1) [inline examples/reason-act.sh:9]: Wouldn’t this always evaluate to `-m -m`?

Comment (tobi) [reply examples/reason-act.sh:9]: err, fixed

**Review by j-f1 — APPROVED** [review]
Cool! At least with the alpaca 7B model I struggle to get it to try an action but it’s a cool idea.

**Comment by sergiobayona** [comment]
great PR. How would I pipe the python statement in `calculate[]`? I assume main.cpp would need to be modified to look for the `Action: ` line then pass it to execute a python `calculate` method. I'm not proficient in CPP enough to do it.

**Pull request merged**